### PR TITLE
Replace tsearch with FI_PEEK + man page updates

### PR DESCRIPTION
**Pull request by shefty**
@jithinjosepkl @j-xiong - I need help updating the PSM and sockets provider to generate asynchronous events for what was tsearch.

### CODE REVIEW

**Comment by shefty** [comment]
Addresses issues #854, #825, #533

File: man/fi_eq.3.md [hunk 1]
```diff
@@ -306,7 +306,7 @@ struct fi_eq_cm_entry {
   If a connection request has been accepted, an FI_CONNECTED event will
   be generated on both sides of the connection.  The active side -- one
   that called fi_connect() -- may receive user data as part of the
-  FI_CONNECTED event.  The user data is passed to the connection
+  FI_CONNECTED event.  The user data is passed to the connectionto
```
Comment (shefty) [inline man/fi_eq.3.md]: gah - not sure where this random 'to' came from.

**Comment by j-xiong** [comment]
I'll update the PSM provider with a separate PR. A question: what is the call to receive the message that has been previously checked and reserved with the FI_PEEK and FI_CLAIM flag?

**Comment by shefty** [comment]
The app uses trecv as before.  It just doesn't specify the FI_PEEK flag.
IOW, trecv(..., FI_PEEK) == tsearch()

**Comment by j-xiong** [comment]
That's doesn't prevent the message from being received by another trecv. It defeats the purpose of FI_CLAIM. There needs to be some mechanism to associate the recv with the message that has been claimed. One way is to return a special context in the completion entry associated with the FI_PEEK|FI_CLAIM flagged recv. And use this special context together with another flag (maybe we can reuse FI_CLAIM here) in a follow up recv to indicate this is a dedicated received for that specific message.

**Comment by sayantansur** [comment]
I agree with Jianxin. The man pages could use some clarification that the 'context' returned with a completion with FI_CLAIM needs to be used to retrieve the message.

**Comment by sayantansur** [comment]
Additionally, we need clarification on behavior of FI_PEEK:

1. How much data is delivered by FI_PEEK operation? Currently, we only have one length parameter in the CQ entry types. How does the app know if the sender wants to send more data than the provided receive buffer? It could look at the length parameter of the completion entry and if > length of posted recv buffer, it can compute the entire length of the message that was sent. If length in completion entry is < length of posted buffer, then the message entirely fit.

2. What happens to data returned by PEEK? Does the data still remain buffered in the provider? I would assume that this remains in the provider until a receive is posted.

3. How big of a receive buffer should the application post to retrieve the remaining buffer? I would assume that the application would need to post a buffer that is large enough to retrieve the data. Otherwise, it can also discard the message using FI_DISCARD. The FI_DISCARD flag should be able to be used with FI_CLAIM as well, otherwise the 'claimed' messages cannot be discarded.

Are these the desired behaviors? It looks like we could use a NOTES section in fi_tagged.3.

**Comment by sayantansur** [comment]
Couple of more things:

1. When FI_PEEK with trecv(buf=NULL), does the fi_cq_tagged_entry::buf point to a provider buffer? if available?

2. What happens when FI_PEEK is used but there's no matching message. Does the receive get posted to the receive queue? It would be nice to control this - I can see folks wanting it both ways :)

**Comment by j-xiong** [comment]
I would suggest to allow receiving directly into the provided buffer if and only if the buffer is large enough to receive the entire message. buffer that is not large enough is treated the same way as NULL and the call would behave the same as the former tsearch with async completion. The completion can use the "flags" field to distinguish between the two case. Partial receiving would require buffering inside the provider,  and that may be contrary to the intention of the FI_PEEK flag.

**Comment by shefty** [comment]
1. The capability to return data as part of a peek request is provider specific.  A provider may always return no data.  If the app does not provide a buffer, the provider would never return data.
2. The amount of data returned should be returned through the 'len' field of the completion.  I was thinking that the olen field would contain the total length - len received.  But olen is also only available for fi_cq_err_entry format.  So an app that wants to peek at partial data would need to use that CQ format or one defined that contains both a len and olen fields.  Some sort of update is needed here.
3. Any data that is peeked is left for a future receive.  The FI_PEEK flag is trying to simulate the MSG_PEEK socket flag, but taking into account asynchronous operation.  That is all.
4. I considered the idea of having an FI_PEEK call post to the receive queue and remain persistent.  I.e. the peek doesn't complete until it finds a match.  The conclusion I reached is that we can add that capability via another flag, if it ever becomes needed by an app.  But the default behavior of peek should be to return 'immediately' with whatever data may be found.

**Comment by sayantansur** [comment]
These sound good to me.

**Comment by shefty** [comment]
I updated the usage for the FI_PEEK, FI_CLAIM, and FI_DISCARD flags.  The behavior wasn't clearly defined in the older man pages, so hopefully this helps and is correct.

**Comment by j-xiong** [comment]
The man page update is very helpful.

**Pull request merged**